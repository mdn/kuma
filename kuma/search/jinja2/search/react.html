{% extends "react_base.html" %}
{% set title = _('Search') %}
{% set query = request.GET.get('q') %}
{% block title %}
  {% if query %}
    {{ page_title(_('Search Results for "%(query)s"', query=query)) }}
  {% else %}
    {{ page_title(_('Search')) }}
  {% endif %}
{% endblock %}

{% block extrahead %}
  {% stylesheet 'react-search' %}
{% endblock %}

{% block robots_value %}noindex, nofollow{% endblock %}

{% block document_head %}
{{ render_react("SPA", request.LANGUAGE_CODE, request.get_full_path(),
                None, False)|safe }}
{% endblock %}
