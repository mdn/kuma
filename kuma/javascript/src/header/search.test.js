//@flow
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import {
    toBeInTheDocument,
    toHaveClass,
} from '@testing-library/jest-dom/matchers';

expect.extend({ toBeInTheDocument, toHaveClass });

import Search from './search.jsx';

const setup = (mockQuery = '') => {
    render(<Search initialQuery={mockQuery} />);

    const form = screen.getByRole('search');
    const input = screen.getByLabelText('Search MDN');
    const openButton = screen.queryByRole('button', {
        name: /open search/i,
    });

    return {
        form,
        input,
        openButton,
    };
};

describe('Search form', () => {
    it('renders default state', () => {
        const { form, openButton } = setup();

        // search form
        expect(form).toBeInTheDocument();

        // open search button
        expect(openButton).toBeInTheDocument();
    });

    it('populates search field with existing query', () => {
        const mockQuery = 'fake search';
        const { input } = setup(mockQuery);
        expect(input.value).toBe(mockQuery);
    });

    it('toggles form styles when button is clicked', () => {
        const { form, openButton } = setup();
        const expectedClassName = 'show-form';

        // no show form class name by default
        expect(form.parentElement).not.toHaveClass(expectedClassName);

        fireEvent.click(openButton);

        const closeButton = screen.queryByRole('button', {
            name: /close search/i,
        });

        // close button shows, open button is gone, classname is applied
        expect(closeButton).toBeInTheDocument;
        expect(openButton).not.toBeInTheDocument;
        expect(form.parentElement).toHaveClass(expectedClassName);

        fireEvent.click(closeButton);

        // default state is restored:
        // close button is gone, open button shows, form has no class name
        expect(closeButton).not.toBeInTheDocument;
        expect(openButton).toBeInTheDocument;
        expect(form.parentElement).not.toHaveClass(expectedClassName);
    });

    it('clears search input when close button is clicked', () => {
        const { openButton, input } = setup();
        const mockValue = 'mock value';

        fireEvent.click(openButton);

        fireEvent.change(input, { target: { value: mockValue } });

        expect(input.value).toBe(mockValue);

        const closeButton = screen.getByRole('button', {
            name: /close search/i,
        });

        fireEvent.click(closeButton);

        expect(input.value).toBe('');
    });
});
