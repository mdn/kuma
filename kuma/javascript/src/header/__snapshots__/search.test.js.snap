// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search snapshot 1`] = `
.emotion-2 {
  grid-area: S;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 2px 8px;
  border: solid 2px #333;
  border-radius: 20px;
  background-color: #fafafa;
  box-sizing: border-box;
  height: 40px;
  min-width: 80px;
  -webkit-flex: 2 1;
  -ms-flex: 2 1;
  flex: 2 1;
}

.emotion-0 {
  fill: #333;
  -webkit-flex: 0 0 20px;
  -ms-flex: 0 0 20px;
  flex: 0 0 20px;
}

.emotion-1 {
  border-width: 0 !important;
  font-size: 14px;
  -webkit-flex: 1 1 100px;
  -ms-flex: 1 1 100px;
  flex: 1 1 100px;
  min-width: 60px;
  max-width: 320px;
  color: #000;
  background-color: #fafafa !important;
}

<form
  action="/en-US/search"
  className="header-search emotion-2"
  id="nav-main-search"
  method="get"
  role="search"
>
  <svg mock
    className="emotion-0"
  />
  <label
    className="visually-hidden"
    htmlFor="main-q"
  >
    Search MDN
  </label>
  <input
    className="emotion-1"
    id="main-q"
    name="q"
    placeholder="Search MDN"
    type="search"
  />
</form>
`;
